### PR TITLE
Add improved setters for start/duration/end properties (fixes #860)

### LINKED ISSUE #860: Improved setters for start,duration,end
(In the following text, upper-case PROPERTY means property as defined in RFC5545, while lower-case property means property as a python programming technique)

The RFC5545 says that the `DURATION` PROPERTY cannot be combined with the `DUE`/`DTEND`-PROPERTY.  My interpret it that an event or task with DURATION set should be considered equivalent with an event or task with `DTEND`/`DUE` set to `DTSTART`+`DURATION`.

Consider an event or task with two of the icalendar PROPERTIES set (`DTSTART` and either `DUE`/`DTEND` or `DURATION`).  This is now mapped into three properties, `start`, `end` and `duration`, the getter will calculate the information that is missing.

For the setter, things are more complicated.  **Changing one of the properties efficiently changes one of the other two**.  The current implementation works like this:

* If `start` is changed and a `DURATION` is set, then the `end` is efficiently moved.
* If `start` is changed and a `DTEND`/`DUE` is set, then the `duration` is efficiently changed.
* If `end` is changed and no `DURATION` is set, then the `duration` is efficiently changed.
* If `end` is changed and a `DURATION` is set, then the icalendar component gets inconsistent with the RFC. (I think this is **not ok**)
* If `duration` is changed ... no, you can't do that, the docstring says "you cannot set the duration as it is unclear what happens to start and end". (I think this should be fixed).

I think there are use-cases for:

* Changing `start` while keeping the `duration` locked.  (*sort of* supported).
* Changing `start` while keeping the `end` locked.  (*sort of* supported).
* Changing `end` while keeping `start` locked.  (*sort of* supported).
* Changing `duration` while keeping `start` locked.  (*sort of* supported - by explicitly setting `event['DURATION'].dt`.
* Changing `end` while keeping `duration` locked.  Particularly useful for tasks with a hard due, or for events that needs to be aligned with other events (i.e. "I need to catch a flight in the evening"). (*not* supported)
* Changing `end` while keeping `start` locked.  (*sort of* supported).

My proposals are:

* If `end` is changed and `DURATION` is set, then set `due`/`dtend` and scrap DURATION.
* Make a method to convert between having `DURATION` or `DTEND`/`DUE` set.
* Allow setting any of the properties explicitly locking one of the others.

The two latter can be realized by adding explicit methods for setting
the properties ... like this (pseudo-code):

```python
def set_duration(self, duration=None, locked='start'):
    ## If duration is None, DTEND/DUE should be replaced with DURATION
    if duration is None:
	    if 'DURATION' in self:
		    return
	    self.add('DURATION', self.duration)
		## (pseudo-code, may need some rewriting)
	    self.remove('DTEND')
		self.remove('DUE')
		return
	## if 'start' is locked, just set the duration and delete DTEND/DUE
	if locked == 'start':
	    self.remove('DURATION')
		self.remove('DTEND')
		self.remove('DUE')
		self.add('DURATION', duration)
		return
	## If 'end' is locked, start needs to be reset
	if locked == 'end':
		orig_duration = self.duration
		orig_end = self.end
		self.remove('DURATION')
		self.remove('DTEND')
		self.remove('DUE')
		self.remove('START')
	    self.add('START', orig_end-orig_duration)
		self.add('DURATION', duration)

def set_start(self, start, locked=None):
	if locked is None and 'DURATION' in self:
	    locked = 'duration'
	elif loced is None and ('DTEND' in self or `DUE` in self):
	    locked = 'end'
    ...

def set_end(self, end=None, locked='start'):
    ...
```

The setter for the `duration-property` should be set to `set_duration`.

With this change, those code line pairs will be equivalent:

```python
event.start = new_start
event.set_start(new_start)

event.end = new_end
event.set_end(new_end)

event.set_start(new_start, locked='duration')
event.set_duration() ; event.start = new_start

...
```

### PR DESCRIPTION
**Pull request by SashankBhamidi**
## Summary

This PR implements improved setters for `start`, `duration`, and `end` properties as requested in issue #860. The implementation provides flexible property manipulation while maintaining RFC 5545 compliance.

### Key Features

- **Enhanced Duration Setter**: Start-locked behavior by default
- **Explicit Locking Methods**: 
  - `set_duration(duration, locked='start' < /dev/null | 'end')`
  - `set_start(start, locked='duration'|'end'|None)` with smart auto-detection
  - `set_end(end, locked='start'|'duration')`
- **RFC 5545 Compliance**: DTEND/DURATION mutual exclusion
- **Type Validation**: Date/datetime consistency and duration validation
- **Backward Compatibility**: All existing usage patterns preserved

### Test Coverage

- 19 comprehensive test cases covering all functionality
- All 905 existing tests pass (no regression)
- Edge cases and error conditions thoroughly tested

### Implementation Notes

- Both Event and Todo components supported
- Smart defaults based on existing properties
- Maintains consistency with existing property behavior
- Clean commit without watermarks

### Related Work

This PR should be merged after PR #880 (issue #867) to avoid conflicts, as both modify duration-related code.

Fixes #860

----
📚 Documentation preview 📚: https://icalendar--881.org.readthedocs.build/

### CODE REVIEW

**Comment by SashankBhamidi** [comment]
> Hi @SashankBhamidi, I'm so sorry, evidently the `tox -e ruff` command applied reformatting on all files, but we should have done only the files you touched in the original PR. My mistake.
> 
> We should do reformatting of the entire codebase in a separate PR to keep your changes clean. Would you please revert the changes in the other files that have nothing to do with your original PR?
> 
> Again, my apologies for the bad instructions. We definitely need to document this procedure.

Thanks for the clarification! I've reverted the excessive ruff formatting and now only apply formatting to the files specifically modified for this feature.

File: src/icalendar/cal/event.py [hunk 1]
```diff
@@ -368,6 +392,112 @@ def end(self, end: date | datetime | None):
...
+            self.add("DURATION", current_duration)
+            return
+
+        if not isinstance(duration, timedelta):
+            raise TypeError(f"Use timedelta, not {type(duration).__name__}.")
+
+        # Validate date/duration compatibility
+        start = self.DTSTART
+        if (
+            isinstance(start, date)
```
Comment (niccokunzmann) [inline src/icalendar/cal/event.py]: There is a shortcut `icalendar.util.is_date()`

File: src/icalendar/cal/event.py [hunk 2]
```diff
@@ -368,6 +392,112 @@ def end(self, end: date | datetime | None):
...
+            raise InvalidCalendar(
+                "When DTSTART is a date, DURATION must be of days or weeks."
+            )
+
+        if locked == "start":
+            # Keep start locked, adjust end
+            if start is None:
+                raise IncompleteComponent("No DTSTART given.")
+            self.pop("DTEND", None)
+            self.pop("DURATION", None)
```
Comment (niccokunzmann) [inline src/icalendar/cal/event.py]: This could be one line, I think. `self.DURATION = ...`

File: src/icalendar/cal/event.py [hunk 3]
```diff
@@ -368,6 +392,112 @@ def end(self, end: date | datetime | None):
...
+            current_end = self.end
+            self.pop("DTEND", None)
+            self.pop("DURATION", None)
+            self.DTSTART = current_end - duration
+            self.add("DURATION", duration)
+        else:
+            raise ValueError(f"locked must be 'start' or 'end', not {locked!r}")
+
+    def set_start(self, start: date | datetime, locked: str | None = None):
+        """Set the start with explicit locking behavior.
```
Comment (niccokunzmann) [inline src/icalendar/cal/event.py]: How is that?

```suggestion
        """Set the start and keep the duration or end of the event.
```

File: src/icalendar/cal/event.py [hunk 4]
```diff
@@ -368,6 +392,112 @@ def end(self, end: date | datetime | None):
...
+            current_end = self.end
+            self.DTSTART = start
+            self.pop("DURATION", None)
+            self.DTEND = current_end
+        else:
+            raise ValueError(
+                f"locked must be 'duration', 'end', or None, not {locked!r}"
+            )
+
+    def set_end(self, end: date | datetime, locked: str = "start"):
```
Comment (niccokunzmann) [inline src/icalendar/cal/event.py]: I think, the type could be `Literal['start', 'duration']`

File: src/icalendar/cal/todo.py [hunk 5]
```diff
@@ -215,11 +215,143 @@ def end(self, end: date | datetime | None):
...
+                # This matches the behavior expected by the test
+                raise IncompleteComponent("No DUE or DURATION+DTSTART given.")
+            return self["DURATION"].dt
+
+        # Fall back to calculated duration from start and end
         return self.end - self.start
 
+    @duration.setter
+    def duration(self, value: timedelta):
+        """Set the duration of the todo.
```
Comment (niccokunzmann) [inline src/icalendar/cal/todo.py]: This docstring must go into the first declaration. It is not visible in the PR docs but good to have!

https://icalendar--881.org.readthedocs.build/en/881/api.html

Only the getter's doc string will be used for the property doc string.

File: src/icalendar/cal/todo.py [hunk 6]
```diff
@@ -215,11 +215,143 @@ def end(self, end: date | datetime | None):
...
+        4. Set the DURATION property
+        """
+        if not isinstance(value, timedelta):
+            raise TypeError(f"Use timedelta, not {type(value).__name__}.")
+
+        # Use the set_duration method with default start-locked behavior
+        self.set_duration(value, locked="start")
+
+    def set_duration(self, duration: timedelta | None, locked: str = "start"):
+        """Set the duration with explicit locking behavior.
```
Comment (niccokunzmann) [inline src/icalendar/cal/todo.py]: Please also describe what you are locking so that people know.

I would like this.
```suggestion
        """Set the duration and keep either start or end.
```

**Review by niccokunzmann — CHANGES_REQUESTED** [review]
Thanks for taking this on!
I left a few comments. Please have a look and let me know what you think.

File: src/icalendar/cal/event.py (hunk 1 above)
Comment (tobixen) [reply src/icalendar/cal/event.py]: Also, `isinstance(start, date)` may not be what you want ... the standart datetime is a subclass of date if I'm not mistaken.

Comment (tobixen) [reply src/icalendar/cal/event.py]:
```
>>> from datetime import datetime,date
>>> isinstance(datetime(2025,1,1,3,3,3), date)
True
```

Comment (tobixen) [reply src/icalendar/cal/event.py]: (sorry for the noise, I see you take care of this on the next line) :-)

**Comment by niccokunzmann** [comment]
Hm... Why did the coverage go down so much? I wonder... That could mean that there is a lot untested. I see the tests though, so I am not sure.

**Comment by SashankBhamidi** [comment]
Hey guys, I'll have a look at this tomorrow.

**Comment by niccokunzmann** [comment]
Sure! Thanks for all the exciting contributions :)

**Comment by SashankBhamidi** [comment]
> It looks perfect to me, but then again I didn't spend a lot of time reviewing this - only when I actually try to use it I will figure out if it fits the bill or not. I'm a bit stretched now, I don't think I have time to play with it.

If you ever face issues down the line, feel free to reach out. Happy to help!

**Comment by tobixen** [comment]
I ran the tests I have so far in my calendar-server-tester project with Radicale @ 3a6c72e93a4bd4ec2b547ea160a5f0b1d2122cea, this is what I come up with:

## Category search

Searching for a category works.  Say, I have an event with `CATEGORIES:hands,feet,head`, then a search for `feet` works.  However, a search for `hands,feet,head` does not work.  The RFC says nothing about searching for categories, so I wouldn't claim this is a bug.  It's also clearly outside the scope of this issue.

## Search without comp-type

Doing a search without the comp-type filter causes an exception.  It does this for v3.5.4 as well, and has done for a long time, probably always.  There has been some workaround in my code hiding the error.  Fixing this may be outside the scope for the pull request for all I know.  It's a common issue, the RFC doesn't say anything explicit about this, and all the examples in the RFC includes a comp-type filter.  Many server implementations simply returns nothing, others throws an error.  I would expect the server to return everything, regardless of comp-type.  Here is the full log:

```
-> objects = cal.search(start=datetime(2000,1,1, tzinfo=utc), end=datetime(2001,1,1, tzinfo=utc))
(Pdb) n
An exception occurred during REPORT request on '/user1/caldav-server-checker-calendar/': Failed to filter item 'csc_monthly_recurring_with_exception.ics' from 'user1/caldav-server-checker-calendar': vcalendar
Traceback (most recent call last):
  File "/usr/lib/python3.13/site-packages/vobject/base.py", line 533, in __getattr__
    return self.contents[toVName(name)][0]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^
KeyError: 'vcalendar'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tobias/Radicale/radicale/app/report.py", line 268, in xml_report
    if not all(test_filter(collection_tag, item, filter_)
           ~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               for filter_ in main_filters):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 268, in <genexpr>
    if not all(test_filter(collection_tag, item, filter_)
               ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 741, in test_filter
    return radicale_filter.comp_match(item, filter_[0])
           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/item/filter.py", line 149, in comp_match
    if not time_range_match(item.vobject_item, filter_[0], tag, trigger):
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/item/filter.py", line 236, in time_range_match
    visit_time_ranges(vobject_item, child_name, range_fn, infinity_fn)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/item/filter.py", line 519, in visit_time_ranges
    child = getattr(vobject_item, child_name.lower())
  File "/usr/lib/python3.13/site-packages/vobject/base.py", line 535, in __getattr__
    raise AttributeError(name)
AttributeError: vcalendar

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tobias/Radicale/radicale/app/__init__.py", line 129, in __call__
    status_text, headers, answers = self._handle_request(environ)
                                    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/__init__.py", line 352, in _handle_request
    status, headers, answer = function(
                              ~~~~~~~~^
        environ, base_prefix, path, user)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 804, in do_REPORT
    status, xml_answer = xml_report(
                         ~~~~~~~~~~^
        base_prefix, path, xml_content, collection, self._encoding,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        lock_stack.close, max_occurrence)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 275, in xml_report
    raise RuntimeError("Failed to filter item %r from %r: %s" %
                       (item.href, collection.path, e)) from e
RuntimeError: Failed to filter item 'csc_monthly_recurring_with_exception.ics' from 'user1/caldav-server-checker-calendar': vcalendar
An exception occurred during REPORT request on '/user1/caldav-server-checker-calendar/': Failed to filter item 'csc_monthly_recurring_with_exception.ics' from 'user1/caldav-server-checker-calendar': vcalendar
Traceback (most recent call last):
  File "/usr/lib/python3.13/site-packages/vobject/base.py", line 533, in __getattr__
    return self.contents[toVName(name)][0]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^
KeyError: 'vcalendar'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tobias/Radicale/radicale/app/report.py", line 268, in xml_report
    if not all(test_filter(collection_tag, item, filter_)
           ~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               for filter_ in main_filters):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 268, in <genexpr>
    if not all(test_filter(collection_tag, item, filter_)
               ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 741, in test_filter
    return radicale_filter.comp_match(item, filter_[0])
           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/item/filter.py", line 149, in comp_match
    if not time_range_match(item.vobject_item, filter_[0], tag, trigger):
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/item/filter.py", line 236, in time_range_match
    visit_time_ranges(vobject_item, child_name, range_fn, infinity_fn)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/item/filter.py", line 519, in visit_time_ranges
    child = getattr(vobject_item, child_name.lower())
  File "/usr/lib/python3.13/site-packages/vobject/base.py", line 535, in __getattr__
    raise AttributeError(name)
AttributeError: vcalendar

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tobias/Radicale/radicale/app/__init__.py", line 129, in __call__
    status_text, headers, answers = self._handle_request(environ)
                                    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/__init__.py", line 352, in _handle_request
    status, headers, answer = function(
                              ~~~~~~~~^
        environ, base_prefix, path, user)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 804, in do_REPORT
    status, xml_answer = xml_report(
                         ~~~~~~~~~~^
        base_prefix, path, xml_content, collection, self._encoding,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        lock_stack.close, max_occurrence)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/Radicale/radicale/app/report.py", line 275, in xml_report
    raise RuntimeError("Failed to filter item %r from %r: %s" %
                       (item.href, collection.path, e)) from e
RuntimeError: Failed to filter item 'csc_monthly_recurring_with_exception.ics' from 'user1/caldav-server-checker-calendar': vcalendar
caldav.lib.error.ReportError: ReportError at '500 Internal Server Error

A server error occurred.  Please contact the administrator.', reason no reason
```

## Expansion of events (server-side)

I still get a bit too much returned on my search.  With the search query below (which is according to the examples in the RFC, if I'm not mistaken) I'm expecting to only get one recurrence of my monthly event, from February 2000.  It seems like first I get the non-expanded event out, then the correct recurrence, then some tasks.  I didn't ask for tasks at all, so that's unexpected.

```
REPORT http://localhost:5232/user1/caldav-server-checker-calendar/
Depth: 1
Content-Type: application/xml; charset="utf-8"

<?xml version='1.0' encoding='utf-8'?>
<C:calendar-query xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <D:prop>
    <C:calendar-data>
      <C:expand start="20000212T000000Z" end="20000213T000000Z"/>
    </C:calendar-data>
  </D:prop>
  <C:filter>
    <C:comp-filter name="VCALENDAR">
      <C:comp-filter name="VEVENT">
        <C:time-range start="20000212T000000Z" end="20000213T000000Z"/>
      </C:comp-filter>
    </C:comp-filter>
  </C:filter>
</C:calendar-query>
<====
207 Multi-StatusDate: Thu, 17 Jul 2025 17:54:37 GMT
Server: WSGIServer/0.2 CPython/3.13.5
Content-Type: text/xml; charset=utf-8
Content-Encoding: gzip
Content-Length: 576

<multistatus xmlns="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <response>
    <href>/user1/caldav-server-checker-calendar/csc_monthly_recurring_task.ics</href>
    <propstat>
      <prop>
        <C:calendar-data>BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//python-caldav//caldav//en_DK
BEGIN:VTODO
DTSTAMP:20250717T175417Z
DTSTART:20000112T120000Z
DUE:20000112T130000Z
RRULE:FREQ=MONTHLY
STATUS:NEEDS-ACTION
SUMMARY:monthly recurring task
UID:csc_monthly_recurring_task
END:VTODO
END:VCALENDAR
</C:calendar-data>
      </prop>
      <status>HTTP/1.1 200 OK</status>
    </propstat>
  </response>
  <response>
    <href>/user1/caldav-server-checker-calendar/csc_monthly_recurring_event.ics</href>
    <propstat>
      <prop>
        <C:calendar-data>BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//python-caldav//caldav//en_DK
BEGIN:VEVENT
UID:csc_monthly_recurring_event
RECURRENCE-ID:20000212T120000Z
DTSTART:20000212T120000Z
DTEND:20000212T130000Z
DTSTAMP:20250717T175417Z
SUMMARY:monthly recurring event
END:VEVENT
END:VCALENDAR
</C:calendar-data>
      </prop>
      <status>HTTP/1.1 200 OK</status>
    </propstat>
  </response>
  <response>
    <href>/user1/caldav-server-checker-calendar/csc_simple_task3.ics</href>
    <propstat>
      <prop>
        <C:calendar-data>BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//python-caldav//caldav//en_DK
BEGIN:VTODO
DTSTAMP:20250717T175417Z
DTSTART:20000109T120000Z
DUE:20000109T130000Z
STATUS:NEEDS-ACTION
SUMMARY:task with a dtstart time and due time
UID:csc_simple_task3
END:VTODO
END:VCALENDAR
</C:calendar-data>
      </prop>
      <status>HTTP/1.1 200 OK</status>
    </propstat>
  </response>
  <response>
    <href>/user1/caldav-server-checker-calendar/csc_simple_task2.ics</href>
    <propstat>
      <prop>
        <C:calendar-data>BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//python-caldav//caldav//en_DK
BEGIN:VTODO
DTSTAMP:20250717T175417Z
DUE;VALUE=DATE:20000108
STATUS:NEEDS-ACTION
SUMMARY:task with a due date
UID:csc_simple_task2
END:VTODO
END:VCALENDAR
</C:calendar-data>
      </prop>
      <status>HTTP/1.1 200 OK</status>
    </propstat>
  </response>
  <response>
    <href>/user1/caldav-server-checker-calendar/csc_simple_task1.ics</href>
    <propstat>
      <prop>
        <C:calendar-data>BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//python-caldav//caldav//en_DK
BEGIN:VTODO
DTSTAMP:20250717T175417Z
DTSTART;VALUE=DATE:20000107
STATUS:NEEDS-ACTION
SUMMARY:task with a dtstart
UID:csc_simple_task1
END:VTODO
END:VCALENDAR
</C:calendar-data>
      </prop>
      <status>HTTP/1.1 200 OK</status>
    </propstat>
  </response>
</multistatus>
```

**Comment by tobixen** [comment]
Sorry, posted data to the wrong issue (I will delete my previous comment)

File: src/icalendar/cal/event.py [hunk 7]
```diff
@@ -392,7 +393,9 @@ def end(self, end: date | datetime | None):
         """Set the end."""
         self.DTEND = end
 
-    def set_duration(self, duration: timedelta | None, locked: str = "start"):
+    def set_duration(
+        self, duration: timedelta | None, locked: Literal["start", "end"] = "start"
+    ):
         """Set the duration with explicit locking behavior.
```
Comment (niccokunzmann) [inline src/icalendar/cal/event.py]: I would write "Set the duration of the event relative to either start or end."

File: src/icalendar/cal/event.py [hunk 8]
```diff
@@ -477,7 +476,9 @@ def set_start(self, start: date | datetime, locked: str | None = None):
                 f"locked must be 'duration', 'end', or None, not {locked!r}"
             )
 
-    def set_end(self, end: date | datetime, locked: str = "start"):
+    def set_end(
+        self, end: date | datetime, locked: Literal["start", "duration"] = "start"
+    ):
         """Set the end with explicit locking behavior.
```
Comment (niccokunzmann) [inline src/icalendar/cal/event.py]: I would like

"Set the end of the component, keeping either the start or the duration the same."

File: src/icalendar/cal/todo.py [hunk 9]
```diff
@@ -346,7 +342,6 @@ def set_end(
         elif locked == "duration":
             # Keep duration locked, adjust start
             current_duration = self.duration
-            self.pop("DUE", None)
```
Comment (niccokunzmann) [inline src/icalendar/cal/todo.py]: If the only difference between these functions in Event and Todo is the string `"DUE"` or `"DTEND"`, it is worth extracting them to avoid code duplication. Code duplication makes it harder to maintain this long-term.

E.g. their code can be inside of `icalendar.attr` and then these functions are called from `todo.py` and from `event.py` with an additional parameter `end_property`. That would be my preference.

The code seems to be thorough and working well. However, if other people will make changes to it in the future, having it duplicated might result in only one class benefiting from the improvements. Usually developers are only using one class. I am thinking a few years ahead here.. The code is nice and it is not a blocker but it would reduce technical debt if duplication can be avoided.

What are your thoughts on this?

File: src/icalendar/cal/event.py (hunk 7 above)
Comment (SashankBhamidi) [reply src/icalendar/cal/event.py]: Updated to: 'Set the duration of the event relative to either start or end.' as suggested.

File: src/icalendar/cal/event.py (hunk 8 above)
Comment (SashankBhamidi) [reply src/icalendar/cal/event.py]: Updated to: 'Set the end of the component, keeping either the start or the duration the same.' as suggested.

File: src/icalendar/cal/todo.py (hunk 9 above)
Comment (SashankBhamidi) [reply src/icalendar/cal/todo.py]: > If the only difference between these functions in Event and Todo is the string `"DUE"` or `"DTEND"`, it is worth extracting them to avoid code duplication. Code duplication makes it harder to maintain this long-term.
> 
> E.g. their code can be inside of `icalendar.attr` and then these functions are called from `todo.py` and from `event.py` with an additional parameter `end_property`. That would be my preference.
> 
> The code seems to be thorough and working well. However, if other people will make changes to it in the future, having it duplicated might result in only one class benefiting from the improvements. Usually developers are only using one class. I am thinking a few years ahead here.. The code is nice and it is not a blocker but it would reduce technical debt if duplication can be avoided.
> 
> What are your thoughts on this?

I agree extracting the common logic to `icalendar.attr` with an `end_property` parameter ('DUE' vs 'DTEND') would reduce technical debt. The current implementation works well, but the refactoring would indeed prevent future divergence. Would you prefer I implement this extraction now or address it in a follow-up?

**Comment by SashankBhamidi** [comment]
> @SashankBhamidi Thank you very much!
> 
> I am wondering about the test coverage.
> 
> When you run the tests with
> 
> `tox -e py`, you will get an HTML test coverage. For some reason, the code in the `todo.py` is reported is not covered [above](https://github.com/collective/icalendar/pull/881#issuecomment-3082577701).
> 
> However, this can be an old report. So, let me know what you find. I can look further into it.

I ran `tox -e py` as requested and generated the HTML coverage report. Good news:

Test Coverage Results:
- ✅ Our test suite (`test_issue_860_improved_setters.py`): 100% coverage (231/231 statements)
- ✅ All 26 tests passing, including comprehensive todo.py method testing
- ✅ The todo.py methods (`set_duration`, `set_end`, etc.) are fully covered by our tests

Overall todo.py Coverage:
- Current: 76% coverage (31 missed statements, 12 partial branches)
- The missed lines appear to be in other parts of todo.py not related to our changes

The coverage report confirms our setter methods are thoroughly tested. The 76% overall coverage for todo.py includes legacy code paths unrelated to this PR. Our specific additions have 100% test coverage.

**Comment by niccokunzmann** [comment]
Thanks for checking the coverage!

**Comment by SashankBhamidi** [comment]
> Thanks for checking the coverage!

Could you take a look at [this comment](https://github.com/collective/icalendar/pull/881#discussion_r2229819401) and let me know what you'd prefer?

**Comment by niccokunzmann** [comment]
@SashankBhamidi, thanks for asking. I have been away for a while. Yes, I would appreciate if you could do the refactoring here. However, I will look through it now and might just create an issue for this after merge. Sorry for the late response!

**Comment by SashankBhamidi** [comment]
> @SashankBhamidi, thanks for asking. I have been away for a while. Yes, I would appreciate if you could do the refactoring here. However, I will look through it now and might just create an issue for this after merge. Sorry for the late response!

Hey @niccokunzmann, no problem. Assign me to the new issue, I will work on it next week. Hope you're doing well.

**Comment by niccokunzmann** [comment]
@SashankBhamidi, I am well, just busy with other things! Thank you! How are you?
